File: agent_matchmaking/ucp.py
"""Unified Capability Profile — create, validate, and convert UCPs.

Handles interoperability with A2A Agent Cards, MCP tool manifests,
and OpenClaw SKILL.md specs (Section 5.3).
"""

from typing import Any, Dict, List, Optional

from .schema import (
    CAPABILITY_DOMAINS,
    Availability,
    Capability,
    Capacity,
    ComplexityRange,
    Cost,
    CostRate,
    Identity,
    Performance,
    QualityMetrics,
    RegistryListing,
    ReliabilityMetrics,
    SpeedMetrics,
    TaxonomyCodes,
    UnifiedCapabilityProfile,
    _gen_id,
    _now_iso,
)


class UCPBuilder:
    """Fluent builder for constructing UCPs."""

    def __init__(self) -> None:
        self._identity = Identity()
        self._capabilities: List[Capability] = []
        self._performance = Performance()
        self._cost = Cost()
        self._availability = Availability()
        self._extensions: Dict[str, Any] = {}
        self._trust_tier = "declared"

    def identity(
        self,
        amp_id: str = "",
        a2a_card: str = "",
        coc_chain_id: str = "",
        did: str = "",
    ) -> "UCPBuilder":
        self._identity = Identity(
            amp_id=amp_id, a2a_card=a2a_card,
            coc_chain_id=coc_chain_id, did=did,
        )
        return self

    def add_registry(self, registry_type: str, listing_id: str) -> "UCPBuilder":
        self._identity.registries.append(
            RegistryListing(registry_type=registry_type, listing_id=listing_id)
        )
        return self

    def add_capability(
        self,
        domain: str,
        subdomain: str = "",
        description: str = "",
        input_modalities: Optional[List[str]] = None,
        output_modalities: Optional[List[str]] = None,
        tools_used: Optional[List[str]] = None,
        amp_capability_code: str = "",
    ) -> "UCPBuilder":
        self._capabilities.append(Capability(
            domain=domain,
            subdomain=subdomain,
            description=description,
            input_modalities=input_modalities or [],
            output_modalities=output_modalities or [],
            tools_used=tools_used or [],
            taxonomy_codes=TaxonomyCodes(amp_capability=amp_capability_code),
        ))
        return self

    def performance(
        self,
        arp_composite: float = 0.0,
        asa_completion_rate: float = 0.0,
        asa_sample_size: int = 0,
        median_response_ms: int = 0,
        p95_response_ms: int = 0,
        throughput_per_hour: int = 0,
    ) -> "UCPBuilder":
        self._performance = Performance(
            reliability=ReliabilityMetrics(
                asa_completion_rate=asa_completion_rate,
                asa_sample_size=asa_sample_size,
            ),
            quality=QualityMetrics(arp_composite_score=arp_composite),
            speed=SpeedMetrics(
                median_response_time_ms=median_response_ms,
                p95_response_time_ms=p95_response_ms,
                throughput_tasks_per_hour=throughput_per_hour,
            ),
        )
        return self

    def cost(
        self,
        pricing_model: str = "posted_price",
        base_amount: float = 0.0,
        base_per: str = "request",
        currency: str = "USD",
        supports_negotiation: bool = False,
        supports_auction: bool = False,
        payment_rails: Optional[List[str]] = None,
    ) -> "UCPBuilder":
        self._cost = Cost(
            pricing_model=pricing_model,
            base_rate=CostRate(amount=base_amount, currency=currency, per=base_per),
            supports_negotiation=supports_negotiation,
            supports_auction=supports_auction,
            payment_rails=payment_rails or [],
        )
        return self

    def availability(
        self,
        status: str = "active",
        lifecycle_stage: str = "operational",
        current_load_pct: int = 0,
        max_concurrent: int = 1,
    ) -> "UCPBuilder":
        self._availability = Availability(
            status=status,
            alp_lifecycle_stage=lifecycle_stage,
            capacity=Capacity(
                current_load_pct=current_load_pct,
                max_concurrent_tasks=max_concurrent,
            ),
        )
        return self

    def trust_tier(self, tier: str) -> "UCPBuilder":
        self._trust_tier = tier
        return self

    def extension(self, key: str, value: Any) -> "UCPBuilder":
        self._extensions[key] = value
        return self

    def build(self) -> UnifiedCapabilityProfile:
        return UnifiedCapabilityProfile(
            identity=self._identity,
            capabilities=self._capabilities,
            performance=self._performance,
            cost=self._cost,
            availability=self._availability,
            extensions=self._extensions,
            trust_tier=self._trust_tier,
        )


# ---------------------------------------------------------------------------
# Validation
# ---------------------------------------------------------------------------

class UCPValidationError(Exception):
    pass


def validate_ucp(ucp: UnifiedCapabilityProfile) -> List[str]:
    """Validate a UCP. Returns list of warnings (empty = valid)."""
    warnings: List[str] = []

    if not ucp.identity.amp_id:
        warnings.append("identity.amp_id is required")
    if not ucp.capabilities:
        warnings.append("at least one capability is required")

    for i, cap in enumerate(ucp.capabilities):
        if not cap.domain:
            warnings.append(f"capabilities[{i}].domain is required")
        elif cap.domain not in CAPABILITY_DOMAINS:
            warnings.append(
                f"capabilities[{i}].domain '{cap.domain}' not in standard domains "
                f"({', '.join(CAPABILITY_DOMAINS)})"
            )
        if not cap.description:
            warnings.append(f"capabilities[{i}].description is recommended")

    if ucp.trust_tier not in ("declared", "attested", "measured", "verified"):
        warnings.append(f"trust_tier '{ucp.trust_tier}' is not a recognized tier")

    return warnings


# ---------------------------------------------------------------------------
# Converters — A2A Agent Card → UCP (Section 5.3)
# ---------------------------------------------------------------------------

def from_a2a_agent_card(card: Dict[str, Any]) -> UnifiedCapabilityProfile:
    """Convert an A2A Agent Card (JSON dict) to a UCP.

    A2A Agent Cards typically have: name, description, url, skills[],
    authentication, supportedProtocols.
    """
    identity = Identity(
        a2a_card=card.get("url", ""),
    )

    capabilities = []
    for skill in card.get("skills", []):
        capabilities.append(Capability(
            description=skill.get("description", skill.get("name", "")),
            domain=_infer_domain(skill.get("description", "")),
            subdomain=skill.get("name", ""),
            input_modalities=skill.get("inputModes", []),
            output_modalities=skill.get("outputModes", []),
        ))

    if not capabilities and card.get("description"):
        capabilities.append(Capability(
            description=card["description"],
            domain=_infer_domain(card["description"]),
        ))

    return UnifiedCapabilityProfile(
        identity=identity,
        capabilities=capabilities,
        trust_tier="attested",
        extensions={"source_format": "a2a_agent_card", "original_name": card.get("name", "")},
    )


def from_mcp_manifest(manifest: Dict[str, Any]) -> UnifiedCapabilityProfile:
    """Convert an MCP tool manifest to a UCP.

    MCP manifests list tools with name, description, inputSchema.
    """
    tools = manifest.get("tools", [])

    capabilities = []
    all_tools = []
    for tool in tools:
        all_tools.append(tool.get("name", ""))
        capabilities.append(Capability(
            description=tool.get("description", tool.get("name", "")),
            domain=_infer_domain(tool.get("description", "")),
            tools_used=[tool.get("name", "")],
        ))

    return UnifiedCapabilityProfile(
        identity=Identity(),
        capabilities=capabilities,
        trust_tier="declared",
        extensions={"source_format": "mcp_manifest", "tool_count": len(tools)},
    )


def from_openclaw_skill(skill_yaml: Dict[str, Any], description_md: str = "") -> UnifiedCapabilityProfile:
    """Convert an OpenClaw SKILL.md (parsed YAML frontmatter + body) to a UCP."""
    identity = Identity()
    identity.registries.append(
        RegistryListing(registry_type="clawhub", listing_id=skill_yaml.get("name", ""))
    )

    cap_desc = description_md or skill_yaml.get("description", "")
    capabilities = [Capability(
        description=cap_desc,
        domain=_infer_domain(cap_desc),
        tools_used=skill_yaml.get("required_binaries", []),
    )]

    return UnifiedCapabilityProfile(
        identity=identity,
        capabilities=capabilities,
        trust_tier="attested",
        extensions={"source_format": "openclaw_skill", "skill_name": skill_yaml.get("name", "")},
    )


# ---------------------------------------------------------------------------
# Domain inference helper
# ---------------------------------------------------------------------------

_DOMAIN_KEYWORDS = {
    "research": ["research", "search", "survey", "investigate", "literature", "competitive"],
    "development": ["code", "develop", "build", "program", "software", "api", "frontend", "backend", "deploy"],
    "analysis": ["analy", "data", "financial", "legal", "statistic", "insight", "evaluate"],
    "communication": ["translate", "summar", "write", "email", "chat", "communicat", "document"],
    "operations": ["monitor", "deploy", "automat", "orchestr", "pipeline", "backup", "infra"],
    "creative": ["design", "creat", "art", "image", "video", "music", "generat"],
    "security": ["secur", "audit", "vulnerab", "threat", "pentest", "compliance", "encrypt"],
    "domain_specific": [],
}


def _infer_domain(text: str) -> str:
    """Best-effort domain inference from description text."""
    text_lower = text.lower()
    scores = {}
    for domain, keywords in _DOMAIN_KEYWORDS.items():
        score = sum(1 for kw in keywords if kw in text_lower)
        if score > 0:
            scores[domain] = score
    if scores:
        return max(scores, key=scores.get)  # type: ignore[arg-type]
    return "domain_specific"

File: typescript/src/ucp.ts
import {
  CAPABILITY_DOMAINS,
  Availability,
  Capability,
  Capacity,
  Cost,
  CostRate,
  Identity,
  Performance,
  QualityMetrics,
  RegistryListing,
  ReliabilityMetrics,
  SpeedMetrics,
  TaxonomyCodes,
  UnifiedCapabilityProfile,
} from "./schema";

// ---------------------------------------------------------------------------
// UCPBuilder
// ---------------------------------------------------------------------------

export class UCPBuilder {
  private _identity = new Identity();
  private _capabilities: Capability[] = [];
  private _performance = new Performance();
  private _cost = new Cost();
  private _availability = new Availability();
  private _extensions: Record<string, unknown> = {};
  private _trustTier = "declared";

  identity(opts: {
    ampId?: string;
    a2aCard?: string;
    cocChainId?: string;
    did?: string;
  } = {}): this {
    this._identity = new Identity(opts);
    return this;
  }

  addRegistry(registryType: string, listingId: string): this {
    this._identity.registries.push(new RegistryListing(registryType, listingId));
    return this;
  }

  addCapability(opts: {
    domain: string;
    subdomain?: string;
    description?: string;
    inputModalities?: string[];
    outputModalities?: string[];
    toolsUsed?: string[];
    ampCapabilityCode?: string;
  }): this {
    this._capabilities.push(new Capability({
      domain: opts.domain,
      subdomain: opts.subdomain,
      description: opts.description,
      inputModalities: opts.inputModalities,
      outputModalities: opts.outputModalities,
      toolsUsed: opts.toolsUsed,
      taxonomyCodes: new TaxonomyCodes("", opts.ampCapabilityCode ?? ""),
    }));
    return this;
  }

  performance(opts: {
    arpComposite?: number;
    asaCompletionRate?: number;
    asaSampleSize?: number;
    medianResponseMs?: number;
    p95ResponseMs?: number;
    throughputPerHour?: number;
  } = {}): this {
    this._performance = new Performance({
      reliability: new ReliabilityMetrics(
        opts.asaCompletionRate ?? 0,
        opts.asaSampleSize ?? 0,
      ),
      quality: new QualityMetrics({ arpCompositeScore: opts.arpComposite ?? 0 }),
      speed: new SpeedMetrics(
        opts.medianResponseMs ?? 0,
        opts.p95ResponseMs ?? 0,
        opts.throughputPerHour ?? 0,
      ),
    });
    return this;
  }

  cost(opts: {
    pricingModel?: string;
    baseAmount?: number;
    basePer?: string;
    currency?: string;
    supportsNegotiation?: boolean;
    supportsAuction?: boolean;
    paymentRails?: string[];
  } = {}): this {
    this._cost = new Cost({
      pricingModel: opts.pricingModel ?? "posted_price",
      baseRate: new CostRate(opts.baseAmount ?? 0, opts.currency ?? "USD", opts.basePer ?? "request"),
      supportsNegotiation: opts.supportsNegotiation ?? false,
      supportsAuction: opts.supportsAuction ?? false,
      paymentRails: opts.paymentRails ?? [],
    });
    return this;
  }

  availability(opts: {
    status?: string;
    lifecycleStage?: string;
    currentLoadPct?: number;
    maxConcurrent?: number;
  } = {}): this {
    this._availability = new Availability(
      opts.status ?? "active",
      opts.lifecycleStage ?? "operational",
      new Capacity(opts.currentLoadPct ?? 0, opts.maxConcurrent ?? 1),
    );
    return this;
  }

  trustTier(tier: string): this {
    this._trustTier = tier;
    return this;
  }

  extension(key: string, value: unknown): this {
    this._extensions[key] = value;
    return this;
  }

  build(): UnifiedCapabilityProfile {
    return new UnifiedCapabilityProfile({
      identity: this._identity,
      capabilities: this._capabilities,
      performance: this._performance,
      cost: this._cost,
      availability: this._availability,
      extensions: this._extensions,
      trustTier: this._trustTier,
    });
  }
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

export class UCPValidationError extends Error {
  constructor(message: string) { super(message); this.name = "UCPValidationError"; }
}

export function validateUcp(ucp: UnifiedCapabilityProfile): string[] {
  const warnings: string[] = [];

  if (!ucp.identity.ampId) warnings.push("identity.amp_id is required");
  if (ucp.capabilities.length === 0) warnings.push("at least one capability is required");

  for (let i = 0; i < ucp.capabilities.length; i++) {
    const cap = ucp.capabilities[i];
    if (!cap.domain) {
      warnings.push(`capabilities[${i}].domain is required`);
    } else if (!CAPABILITY_DOMAINS.includes(cap.domain as any)) {
      warnings.push(
        `capabilities[${i}].domain '${cap.domain}' not in standard domains (${CAPABILITY_DOMAINS.join(", ")})`,
      );
    }
    if (!cap.description) {
      warnings.push(`capabilities[${i}].description is recommended`);
    }
  }

  if (!["declared", "attested", "measured", "verified"].includes(ucp.trustTier)) {
    warnings.push(`trust_tier '${ucp.trustTier}' is not a recognized tier`);
  }

  return warnings;
}

// ---------------------------------------------------------------------------
// Converters — A2A Agent Card -> UCP (Section 5.3)
// ---------------------------------------------------------------------------

export function fromA2aAgentCard(card: Record<string, unknown>): UnifiedCapabilityProfile {
  const identity = new Identity({ a2aCard: (card.url as string) ?? "" });

  const capabilities: Capability[] = [];
  const skills = (card.skills ?? []) as Array<Record<string, unknown>>;
  for (const skill of skills) {
    capabilities.push(new Capability({
      description: (skill.description as string) ?? (skill.name as string) ?? "",
      domain: inferDomain((skill.description as string) ?? ""),
      subdomain: (skill.name as string) ?? "",
      inputModalities: (skill.inputModes as string[]) ?? [],
      outputModalities: (skill.outputModes as string[]) ?? [],
    }));
  }

  if (capabilities.length === 0 && card.description) {
    capabilities.push(new Capability({
      description: card.description as string,
      domain: inferDomain(card.description as string),
    }));
  }

  return new UnifiedCapabilityProfile({
    identity,
    capabilities,
    trustTier: "attested",
    extensions: { source_format: "a2a_agent_card", original_name: (card.name as string) ?? "" },
  });
}

export function fromMcpManifest(manifest: Record<string, unknown>): UnifiedCapabilityProfile {
  const tools = (manifest.tools ?? []) as Array<Record<string, unknown>>;

  const capabilities: Capability[] = [];
  for (const tool of tools) {
    capabilities.push(new Capability({
      description: (tool.description as string) ?? (tool.name as string) ?? "",
      domain: inferDomain((tool.description as string) ?? ""),
      toolsUsed: [(tool.name as string) ?? ""],
    }));
  }

  return new UnifiedCapabilityProfile({
    identity: new Identity(),
    capabilities,
    trustTier: "declared",
    extensions: { source_format: "mcp_manifest", tool_count: tools.length },
  });
}

export function fromOpenclawSkill(
  skillYaml: Record<string, unknown>,
  descriptionMd = "",
): UnifiedCapabilityProfile {
  const identity = new Identity();
  identity.registries.push(
    new RegistryListing("clawhub", (skillYaml.name as string) ?? ""),
  );

  const capDesc = descriptionMd || ((skillYaml.description as string) ?? "");
  const capabilities = [new Capability({
    description: capDesc,
    domain: inferDomain(capDesc),
    toolsUsed: (skillYaml.required_binaries as string[]) ?? [],
  })];

  return new UnifiedCapabilityProfile({
    identity,
    capabilities,
    trustTier: "attested",
    extensions: { source_format: "openclaw_skill", skill_name: (skillYaml.name as string) ?? "" },
  });
}

// ---------------------------------------------------------------------------
// Domain inference helper
// ---------------------------------------------------------------------------

const DOMAIN_KEYWORDS: Record<string, string[]> = {
  research: ["research", "search", "survey", "investigate", "literature", "competitive"],
  development: ["code", "develop", "build", "program", "software", "api", "frontend", "backend", "deploy"],
  analysis: ["analy", "data", "financial", "legal", "statistic", "insight", "evaluate"],
  communication: ["translate", "summar", "write", "email", "chat", "communicat", "document"],
  operations: ["monitor", "deploy", "automat", "orchestr", "pipeline", "backup", "infra"],
  creative: ["design", "creat", "art", "image", "video", "music", "generat"],
  security: ["secur", "audit", "vulnerab", "threat", "pentest", "compliance", "encrypt"],
  domain_specific: [],
};

export function inferDomain(text: string): string {
  const textLower = text.toLowerCase();
  const scores: Record<string, number> = {};
  for (const [domain, keywords] of Object.entries(DOMAIN_KEYWORDS)) {
    let score = 0;
    for (const kw of keywords) {
      if (textLower.includes(kw)) score++;
    }
    if (score > 0) scores[domain] = score;
  }
  if (Object.keys(scores).length > 0) {
    return Object.entries(scores).sort((a, b) => b[1] - a[1])[0][0];
  }
  return "domain_specific";
}
